File: backend/omninest-media/src/main/java/com/omninest/modules/reader/domain/ReaderProgress.java
package com.omninest.modules.reader.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "reader_progress", schema = "omni")
@Data
@AllArgsConstructor
@NoArgsConstructor
public class ReaderProgress {
    @Id
    private UUID id;

    @Column(name = "owner_user_id", nullable = false)
    private UUID ownerUserId;

    @Column(name = "reader_item_id", nullable = false)
    private UUID readerItemId;

    @Column(name = "char_offset", nullable = false)
    private long charOffset = 0;

    @Column(name = "progress_percent", nullable = false)
    private BigDecimal progressPercent = BigDecimal.ZERO;

    @Column(name = "reading_mode", nullable = false, length = 50)
    private String readingMode = "scroll";

    @Column(name = "chapter_id", nullable = false, length = 128)
    private String chapterId = "";

    /** 漫画当前页面 ID */
    @Column(name = "page_id")
    private UUID pageId;

    /** 漫画当前页面索引 */
    @Column(name = "page_index")
    private Integer pageIndex;

    /** 漫画当前页面指纹（用于校验页面未变更） */
    @Column(name = "page_fingerprint", length = 64)
    private String pageFingerprint;

    /** 漫画当前来源文件 ID */
    @Column(name = "source_id")
    private UUID sourceId;

    /** 漫画当前页面在来源文件内的索引 */
    @Column(name = "source_page_index")
    private Integer sourcePageIndex;

    /** 漫画当前页面所属目录键 */
    @Column(name = "catalog_key", length = 500)
    private String catalogKey;

    /** 漫画：页内偏移（滚动模式，0.0-1.0） */
    @Column(name = "intra_page_offset")
    private Double intraPageOffset;

    /** 漫画目录清单版本号（用于检测目录结构变更） */
    @Column(name = "manifest_version")
    private Integer manifestVersion = 0;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @Version
    @Column(nullable = false)
    private long version;

    @PrePersist
    void fillId() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (updatedAt == null) {
            updatedAt = Instant.now();
        }
    }

    @PreUpdate
    void fillUpdatedAt() {
        updatedAt = Instant.now();
    }
}

File: frontend/lib/features/reader/domain/reader_progress.dart
class ReaderProgress {
  const ReaderProgress({
    required this.readerItemId,
    required this.charOffset,
    required this.progressPercent,
    required this.readingMode,
    required this.updatedAt,
    this.chapterId,
    this.pageId,
    this.pageIndex,
    this.pageFingerprint,
    this.sourceId,
    this.sourcePageIndex,
    this.catalogKey,
    this.manifestVersion,
    this.intraPageOffset,
  });

  factory ReaderProgress.fromJson(Map<String, dynamic> json) {
    return ReaderProgress(
      readerItemId: json['readerItemId']?.toString() ?? '',
      charOffset: _asInt(json['charOffset']),
      progressPercent: _asDouble(json['progressPercent']),
      readingMode: json['readingMode']?.toString() ?? 'scroll',
      chapterId: json['chapterId']?.toString(),
      updatedAt: _parseDateTime(json['updatedAt']),
      pageId: json['pageId']?.toString(),
      pageIndex: json['pageIndex'] as int?,
      pageFingerprint: json['pageFingerprint']?.toString(),
      sourceId: json['sourceId']?.toString(),
      sourcePageIndex: json['sourcePageIndex'] as int?,
      catalogKey: json['catalogKey']?.toString(),
      manifestVersion: json['manifestVersion'] as int?,
      intraPageOffset: _asDoubleNullable(json['intraPageOffset']),
    );
  }

  final String readerItemId;

  /// 章节内字符偏移，与 chapterId 共同构成精确位置。
  final int charOffset;

  /// 全书进度比例（0-1），仅用于 UI 显示，不参与位置恢复。
  final double progressPercent;
  final String readingMode;
  final String? chapterId;
  final DateTime? updatedAt;

  /// 漫画页面 ID（用于精确定位）
  final String? pageId;

  /// 漫画页面索引
  final int? pageIndex;

  /// 漫画页面指纹（防篡改校验）
  final String? pageFingerprint;

  /// 漫画来源文件 ID
  final String? sourceId;

  /// 漫画页面在来源文件内的索引
  final int? sourcePageIndex;

  /// 漫画目录键
  final String? catalogKey;

  /// 漫画清单版本号
  final int? manifestVersion;

  /// 漫画：页内偏移（滚动模式，0.0-1.0）
  final double? intraPageOffset;

  /// 阅读模式别名
  String get mode => readingMode;
}

int _asInt(dynamic value) => switch (value) {
  int() => value,
  num() => value.toInt(),
  _ => int.tryParse(value?.toString() ?? '') ?? 0,
};

double _asDouble(dynamic value) => switch (value) {
  double() => value,
  num() => value.toDouble(),
  _ => double.tryParse(value?.toString() ?? '') ?? 0.0,
};

double? _asDoubleNullable(dynamic value) {
  if (value == null) return null;
  return switch (value) {
    double() => value,
    num() => value.toDouble(),
    _ => double.tryParse(value.toString()),
  };
}

DateTime? _parseDateTime(dynamic value) {
  if (value == null || value.toString().isEmpty) return null;
  return DateTime.tryParse(value.toString().replaceFirst(' ', 'T'))?.toLocal();
}
